### PR TITLE
Update CycloneDX to use syft namespace and output multiple CPEs

### PR DESCRIPTION
**Pull request by kzantow**
This fixes: #818 and #842 and will complete the encoder/decoder work for https://github.com/anchore/grype/issues/481

### CODE REVIEW

File: internal/formats/common/cyclonedxhelpers/format.go [hunk 1]
```diff
@@ -137,7 +139,7 @@ func toDependencies(relationships []artifact.Relationship) []cyclonedx.Dependenc
 	for _, r := range relationships {
 		exists := isExpressiblePackageRelationship(r.Type)
 		if !exists {
-			log.Warnf("unable to convert relationship from CycloneDX 1.3 JSON, dropping: %+v", r)
+			log.Debugf("unable to convert relationship from CycloneDX 1.3 JSON, dropping: %+v", r)
```
Comment (kzantow) [inline internal/formats/common/cyclonedxhelpers/format.go:142]: Not sure this was needed or ideal, but the logs were filled with this message.

File: internal/formats/common/property_encoder.go [hunk 2]
```diff
@@ -0,0 +1,325 @@
+package common
```
Comment (kzantow) [inline internal/formats/common/property_encoder.go:1]: Fun fact: if you take this property_encoder and it's test out of the lines changed, this is a net reduction of code.

File: syft/source/location.go [hunk 3]
```diff
@@ -11,7 +11,7 @@ import (
 // Location represents a path relative to a particular filesystem resolved to a specific file.Reference. This struct is used as a key
 // in content fetching to uniquely identify a file relative to a request (the VirtualPath).
 type Location struct {
-	Coordinates
+	Coordinates `cyclonedx:""`
```
Comment (kzantow) [inline syft/source/location.go:14]: Empty string here causes the sub-properties to be included, without an intermediate property name segment, like `syft:location:0:coordinates:path`

File: test/integration/encode_decode_cycle_test.go [hunk 4]
```diff
@@ -63,10 +66,16 @@ func TestEncodeDecodeEncodeCycleComparison(t *testing.T) {
...
 			}
 
-			if !assert.True(t, bytes.Equal(by1, by2)) {
-				dmp := diffmatchpatch.New()
-				diffs := dmp.DiffMain(string(by1), string(by2), true)
-				t.Errorf("diff: %s", dmp.DiffPrettyText(diffs))
+			if test.json {
+				s1 := string(by1)
+				s2 := string(by2)
+				assert.JSONEq(t, s1, s2)
```
Comment (kzantow) [inline test/integration/encode_decode_cycle_test.go:72]: This is much more useful to figure out what's changed in the json! Great suggestion @wagoodman

File: internal/formats/common/property_encoder.go [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+
+func encode(out map[string]string, value reflect.Value, prefix string, fn FieldName) {
+	if !value.IsValid() || value.Type() == nil {
+		return
+	}
+
+	typ := value.Type()
+
+	switch typ.Kind() {
+	// case reflect.Interface:
```
Comment (wagoodman) [inline internal/formats/common/property_encoder.go:96]: nit: dead code comments could get removed

File: internal/formats/common/property_encoder_test.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
+package common
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+type T1 struct {
```
Comment (wagoodman) [inline internal/formats/common/property_encoder_test.go:10]: ❤️ love the tests here, nice job

File: syft/source/location.go (hunk 3 above)
Comment (wagoodman) [reply syft/source/location.go:14]: this should get captured as a comment on this field in the code for certain

**Review by wagoodman — APPROVED** [review]
all sorts of reflect-fun! Took a while for me to get through it. Nice job 🙌

File: internal/formats/common/cyclonedxhelpers/format.go (hunk 1 above)
Comment (wagoodman) [reply internal/formats/common/cyclonedxhelpers/format.go:142]: this sounds like it should stay a warning. Is the case that it's hitting common / legitimate?

Comment (kzantow) [reply internal/formats/common/cyclonedxhelpers/format.go:142]: it seems to happen a lot including the tests

**Comment by jkowalleck** [comment]
I wonder whether the `syft:` CycloneDX taxonomy has been published. 
If so, could you link it in the registry? 
see https://github.com/CycloneDX/cyclonedx-property-taxonomy/blob/3edec8ccbb8ca63dc9949a695d119cf9f3d51446/README.md?plain=1#L116

**Pull request merged**